{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### All Imports"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from General.args import get_args\n",
    "import General.utils as GU\n",
    "from PD_var_ILP.run_var_ilp import run_var_ilp\n",
    "from PD_mul_ILP.run_mul_ilp import run_mul_ilp\n",
    "from PD_mul_greedy.run_mul_greedy import run_mul_greedy\n",
    "from PD_single_LPath.run_pd_single_LPath import run_longest_path"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Set up gurobi optimization library\n",
    "\n",
    "\n",
    "First, create a file called gurobi.json with the details of your gurobi license in the main directory. The file should be in the following format:\n",
    "\n",
    "```json\n",
    "{\n",
    "  \"WLSACCESSID\": \"XXXXX\",\n",
    "  \"WLSSECRET\": \"XXXXX\",\n",
    "  \"LICENSEID\": 12345\n",
    "}\n",
    "```\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Global program cmd arguments\n",
    "\n",
    "Set arguments to their default values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "args = get_args()\n",
    "\n",
    "args.output = 'Temp'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Define global variables for upstream and downstream regions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "cfg = GU.load_config(\"config.json\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Run PD-Single-LPath"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Define sequence\n",
    "First, we will define our mutreg region sequence, we will take the first 100 nucleotides of the coding sequence of the CXDAR human protein gene as a small test dataset."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "protein_name = 'CXDAR'\n",
    "mutreg_nt = 'ATGGCGCTCCTGCTGTGCTTCGTGCTCCTGTGCGGAGTAGTGGATTTCGCCAGAAGTTTGAGTATCACTACTCCTGAAGAGATGATTGAAAAAGCCAAAGCTACTCCTGAAGAGATGATTGAAAAAGCCAAAG'\n",
    "\n",
    "## Add upstream and downstream sequences\n",
    "full_sequence = cfg.upstream + mutreg_nt + cfg.downstream"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Run longest path algorithm\n",
    "\n",
    "We will runs PD-Single_LPath which runs a longest path algorithms on the primer graph, which is a DAG, to find the most efficient primers."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Saved selected primers to: Temp/PD_single_LPath_selected_primers.csv\n",
      "Results DataFrame:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>protein_name</th>\n",
       "      <th>graph_nodes</th>\n",
       "      <th>graph_edges</th>\n",
       "      <th>graph_time_sec</th>\n",
       "      <th>longest_path_efficiency</th>\n",
       "      <th>total_time_sec</th>\n",
       "      <th>num_primers</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CXDAR</td>\n",
       "      <td>3408</td>\n",
       "      <td>236389</td>\n",
       "      <td>2.949</td>\n",
       "      <td>1.61325</td>\n",
       "      <td>8.372</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  protein_name  graph_nodes  graph_edges  graph_time_sec  \\\n",
       "0        CXDAR         3408       236389           2.949   \n",
       "\n",
       "   longest_path_efficiency  total_time_sec  num_primers  \n",
       "0                  1.61325           8.372            4  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "run_df, primer_paths = run_longest_path(full_sequence, mutreg_nt, protein_name, args, cfg) \n",
    "\n",
    "print(\"Results DataFrame:\")\n",
    "run_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Selected primer path and primer efficiency\n",
    "\n",
    "Print all selected primers in longest path and their efficiency"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Primer Paths:\n",
      "(-58, -34, 'f')\n",
      "(119, 141, 'r')\n",
      "(93, 115, 'f')\n",
      "(277, 296, 'r')\n",
      "Primer effiency\n",
      "1.6132498038728875\n"
     ]
    }
   ],
   "source": [
    "print(\"Primer Paths:\")\n",
    "for path in primer_paths:\n",
    "    print(path)\n",
    "\n",
    "print(\"Primer effiency\")\n",
    "print(run_df['longest_path_efficiency'].prod())   "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Run with Tm & GC content thresholds\n",
    "\n",
    "Apply thresholds on primer melting temperatures (tm), GC content, and maximum difference between fwd and reverse primers "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Saved selected primers to: Temp/PD_single_LPath_selected_primers.csv\n",
      "Primer Paths:\n",
      "(-58, -34, 'f')\n",
      "(123, 143, 'r')\n",
      "(94, 118, 'f')\n",
      "(280, 299, 'r')\n",
      "1.1101513634330775\n"
     ]
    }
   ],
   "source": [
    "args.apply_threshold = True\n",
    "\n",
    "args.min_tm = 58\n",
    "args.max_tm= 65\n",
    "\n",
    "args.min_tm = 40\n",
    "args.max_tm= 60\n",
    "\n",
    "args.max_difference = 3\n",
    "\n",
    "run_df, primer_paths = run_longest_path(full_sequence, mutreg_nt, protein_name, args, cfg) \n",
    "\n",
    "print(\"Primer Paths:\")\n",
    "for path in primer_paths:\n",
    "    print(path)\n",
    "    \n",
    "print(run_df['longest_path_efficiency'].prod())   \n",
    "\n",
    "args.apply_threshold = False"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Run PD-var-ILP\n",
    "\n",
    "We will run PD-var-ILP on multiple variants of the same protein.\n",
    "\n",
    "For this part, we will use the same sequence of the CXDAR protein as PD-single-LPath.\n",
    "We will use the default number of variants, which is 3, and the deafult maximum overlap length, which is 6. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Running greedy algorithm\n",
      "Running ILP\n",
      "Set parameter WLSAccessID\n",
      "Set parameter WLSSecret\n",
      "Set parameter LicenseID to value 2499075\n",
      "Academic license 2499075 - for non-commercial use only - registered to jo___@live.biu.ac.il\n",
      "Number of Constraints: 7104\n",
      "Average Vars Per Constraint 84.39386261261261\n",
      "Graph edges:  236389\n",
      "Finished ILP Variable Creations\n",
      "0\n",
      "3\n",
      "7\n",
      "11\n",
      "15\n",
      "19\n",
      "23\n",
      "27\n",
      "31\n",
      "35\n",
      "39\n",
      "43\n",
      "47\n",
      "51\n",
      "55\n",
      "59\n",
      "63\n",
      "67\n",
      "71\n",
      "75\n",
      "79\n",
      "83\n",
      "87\n",
      "91\n",
      "95\n",
      "99\n",
      "Finished Intersection constraints!\n",
      "Gurobi Optimizer version 12.0.3 build v12.0.3rc0 (linux64 - \"Rocky Linux 9.7 (Blue Onyx)\")\n",
      "\n",
      "CPU model: Intel(R) Xeon(R) Gold 6338 CPU @ 2.00GHz, instruction set [SSE2|AVX|AVX2|AVX512]\n",
      "Thread count: 64 physical cores, 128 logical processors, using up to 32 threads\n",
      "\n",
      "Academic license 2499075 - for non-commercial use only - registered to jo___@live.biu.ac.il\n",
      "Optimize a model with 10512 rows, 236389 columns and 40624023 nonzeros\n",
      "Model fingerprint: 0x309eda85\n",
      "Variable types: 0 continuous, 236389 integer (236389 binary)\n",
      "Coefficient statistics:\n",
      "  Matrix range     [1e+00, 1e+00]\n",
      "  Objective range  [8e-05, 1e+00]\n",
      "  Bounds range     [1e+00, 1e+00]\n",
      "  RHS range        [1e+00, 3e+00]\n",
      "Presolve removed 159 rows and 26 columns (presolve time = 6s)...\n",
      "Presolve removed 159 rows and 26 columns (presolve time = 11s)...\n",
      "Presolve removed 1084 rows and 26 columns (presolve time = 15s)...\n",
      "Presolve removed 1084 rows and 49 columns (presolve time = 20s)...\n",
      "Presolve removed 1084 rows and 49 columns (presolve time = 25s)...\n",
      "Presolve removed 1084 rows and 49 columns (presolve time = 30s)...\n",
      "Presolve removed 1119 rows and 49 columns (presolve time = 35s)...\n",
      "Presolve removed 1119 rows and 49 columns (presolve time = 40s)...\n",
      "Presolve removed 5314 rows and 49 columns (presolve time = 46s)...\n",
      "Presolve removed 5314 rows and 49 columns (presolve time = 51s)...\n",
      "Presolve removed 5314 rows and 49 columns (presolve time = 55s)...\n",
      "Presolve removed 5430 rows and 107 columns (presolve time = 60s)...\n",
      "Presolve removed 5430 rows and 107 columns (presolve time = 65s)...\n",
      "Presolve removed 5430 rows and 107 columns (presolve time = 70s)...\n",
      "Presolve removed 5430 rows and 107 columns (presolve time = 75s)...\n",
      "Presolve removed 5430 rows and 107 columns (presolve time = 80s)...\n",
      "Presolve removed 5430 rows and 107 columns (presolve time = 85s)...\n",
      "Presolve removed 5430 rows and 107 columns (presolve time = 90s)...\n",
      "Presolve removed 5430 rows and 107 columns (presolve time = 95s)...\n",
      "Sparsify removed 13666444 nonzeros (94%)\n",
      "Presolve added 0 rows and 772 columns\n",
      "Presolve removed 5480 rows and 0 columns\n",
      "Presolve time: 96.91s\n",
      "Presolved: 5032 rows, 237161 columns, 840899 nonzeros\n",
      "Variable types: 220 continuous, 236941 integer (236468 binary)\n",
      "Deterministic concurrent LP optimizer: primal simplex, dual simplex, and barrier\n",
      "Showing barrier log only...\n",
      "\n",
      "Root barrier log...\n",
      "\n",
      "Ordering time: 0.26s\n",
      "\n",
      "Barrier statistics:\n",
      " AA' NZ     : 3.514e+05\n",
      " Factor NZ  : 1.576e+06 (roughly 100 MB of memory)\n",
      " Factor Ops : 6.896e+08 (less than 1 second per iteration)\n",
      " Threads    : 29\n",
      "\n",
      "                  Objective                Residual\n",
      "Iter       Primal          Dual         Primal    Dual     Compl     Time\n",
      "   0  -7.70271733e+09  3.24470972e+07  7.77e+08 1.87e+00  4.91e+05   100s\n",
      "   1  -2.59004038e+09  1.61345134e+07  2.61e+08 1.16e+00  1.69e+05   100s\n",
      "   2  -8.46220623e+08  6.20476499e+06  8.53e+07 2.05e-01  5.62e+04   100s\n",
      "\n",
      "Barrier performed 2 iterations in 99.76 seconds (141.70 work units)\n",
      "Barrier solve interrupted - model solved by another algorithm\n",
      "\n",
      "Concurrent spin time: 0.34s (can be avoided by choosing Method=3)\n",
      "\n",
      "Solved with dual simplex\n",
      "\n",
      "Root simplex log...\n",
      "\n",
      "Iteration    Objective       Primal Inf.    Dual Inf.      Time\n",
      "    4164    2.5269911e+00   0.000000e+00   0.000000e+00    100s\n",
      "\n",
      "Root relaxation: objective 2.526991e+00, 4164 iterations, 1.07 seconds (0.97 work units)\n",
      "\n",
      "    Nodes    |    Current Node    |     Objective Bounds      |     Work\n",
      " Expl Unexpl |  Obj  Depth IntInf | Incumbent    BestBd   Gap | It/Node Time\n",
      "\n",
      "*    0     0               0       2.5269911    2.52699  0.00%     -   99s\n",
      "\n",
      "Explored 1 nodes (4164 simplex iterations) in 100.02 seconds (141.74 work units)\n",
      "Thread count was 32 (of 128 available processors)\n",
      "\n",
      "Solution count 1: 2.52699 \n",
      "\n",
      "Optimal solution found (tolerance 1.00e-04)\n",
      "Best objective 2.526991111515e+00, best bound 2.526991111515e+00, gap 0.0000%\n",
      "Protein #1 (4)\n",
      "[\"(-81, -60, 'f')\", \"(92, 114, 'r')\", \"(64, 89, 'f')\", \"(241, 259, 'r')\"]\n",
      "\n",
      "Protein #2 (4)\n",
      "[\"(-58, -34, 'f')\", \"(119, 141, 'r')\", \"(93, 115, 'f')\", \"(277, 296, 'r')\"]\n",
      "\n",
      "Protein #3 (2)\n",
      "[\"(-31, -1, 'f')\", \"(156, 174, 'r')\"]\n",
      "\n",
      "Saved summary to: Temp/var_ILP_results.csv\n",
      "Saved ILP selected primers to: Temp/var_ILP_selected_primers_ILP.csv\n",
      "Saved greedy selected primers to: Temp/var_ILP_selected_primers_greedy.csv\n",
      "Variable ILP Results DataFrame:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>protein_name</th>\n",
       "      <th>num_variants</th>\n",
       "      <th>seq_length</th>\n",
       "      <th>graph_nodes</th>\n",
       "      <th>graph_edges</th>\n",
       "      <th>graph_time_sec</th>\n",
       "      <th>greedy_path_length</th>\n",
       "      <th>ilp_num_vars</th>\n",
       "      <th>ilp_path_length</th>\n",
       "      <th>ilp_num_constraints</th>\n",
       "      <th>ilp_setup_time_sec</th>\n",
       "      <th>ilp_optimize_time_sec</th>\n",
       "      <th>ILP_primer_efficiency</th>\n",
       "      <th>ILP_solution_feasibility</th>\n",
       "      <th>greedy_solution_feasibility</th>\n",
       "      <th>greedy_primer_efficiency</th>\n",
       "      <th>greedy_time_sec</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CXDAR</td>\n",
       "      <td>3</td>\n",
       "      <td>133</td>\n",
       "      <td>3408</td>\n",
       "      <td>236389</td>\n",
       "      <td>2.945</td>\n",
       "      <td>10</td>\n",
       "      <td>236389</td>\n",
       "      <td>10</td>\n",
       "      <td>10512</td>\n",
       "      <td>92.367</td>\n",
       "      <td>101.068</td>\n",
       "      <td>2.526991</td>\n",
       "      <td>FEASIBLE</td>\n",
       "      <td>FEASIBLE</td>\n",
       "      <td>2.526991</td>\n",
       "      <td>1.194</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  protein_name  num_variants  seq_length  graph_nodes  graph_edges  \\\n",
       "0        CXDAR             3         133         3408       236389   \n",
       "\n",
       "   graph_time_sec  greedy_path_length  ilp_num_vars  ilp_path_length  \\\n",
       "0           2.945                  10        236389               10   \n",
       "\n",
       "   ilp_num_constraints  ilp_setup_time_sec  ilp_optimize_time_sec  \\\n",
       "0                10512              92.367                101.068   \n",
       "\n",
       "   ILP_primer_efficiency ILP_solution_feasibility greedy_solution_feasibility  \\\n",
       "0               2.526991                 FEASIBLE                    FEASIBLE   \n",
       "\n",
       "   greedy_primer_efficiency  greedy_time_sec  \n",
       "0                  2.526991            1.194  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "results_df, ilp_path, greedy_path  = run_var_ilp(full_sequence, mutreg_nt, protein_name,args,cfg)\n",
    "print(\"Variable ILP Results DataFrame:\")\n",
    "results_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Compare PD-Var-ILP with greedy baseline\n",
    "\n",
    "We will compare the primers found by the ILP and by the greedy approach and their total efficiencies."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ILP chosen primers:\n",
      "Variant 1\n",
      "[\"(-81, -60, 'f')\", \"(92, 114, 'r')\", \"(64, 89, 'f')\", \"(241, 259, 'r')\"]\n",
      "Variant 2\n",
      "[\"(-58, -34, 'f')\", \"(119, 141, 'r')\", \"(93, 115, 'f')\", \"(277, 296, 'r')\"]\n",
      "Variant 3\n",
      "[\"(-31, -1, 'f')\", \"(156, 174, 'r')\"]\n",
      "ILP primer efficiency:  2.5269911115148025\n",
      "Greedy chosen primers:\n",
      "Variant 1\n",
      "[(-58, -34, 'f'), (119, 141, 'r'), (93, 115, 'f'), (277, 296, 'r')]\n",
      "Variant 2\n",
      "[(-81, -60, 'f'), (92, 114, 'r'), (64, 89, 'f'), (241, 259, 'r')]\n",
      "Variant 3\n",
      "[(-31, -1, 'f'), (156, 174, 'r')]\n",
      "Greedy primer efficiency:  2.526991111514803\n"
     ]
    }
   ],
   "source": [
    "print(\"ILP chosen primers:\")\n",
    "\n",
    "for i,path in enumerate(ilp_path):\n",
    "    print(\"Variant\", i+1)\n",
    "    print(path)\n",
    "\n",
    "print(\"ILP primer efficiency: \",results_df['ILP_primer_efficiency'].prod())\n",
    "\n",
    "\n",
    "print(\"Greedy chosen primers:\")\n",
    "for i,path in enumerate(greedy_path):\n",
    "    print(\"Variant\", i+1)\n",
    "    print(path)\n",
    "\n",
    "print(\"Greedy primer efficiency: \",results_df['greedy_primer_efficiency'].prod())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Run PD-Var_ILP with merging bins corresponding to identical sequences\n",
    "\n",
    "Run using feature which imposes ILP constraints that also merge bins with identical sequence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Running greedy algorithm\n",
      "Running ILP\n",
      "Set parameter WLSAccessID\n",
      "Set parameter WLSSecret\n",
      "Set parameter LicenseID to value 2499075\n",
      "Academic license 2499075 - for non-commercial use only - registered to jo___@live.biu.ac.il\n",
      "Number of Constraints: 7104\n",
      "Average Vars Per Constraint 84.39386261261261\n",
      "Number of united bins constraints: 6596\n",
      "Average Vars Per Constraint 90.89357186173439\n",
      "Graph edges:  236389\n",
      "Finished ILP Variable Creations\n",
      "0\n",
      "3\n",
      "7\n",
      "11\n",
      "15\n",
      "19\n",
      "23\n",
      "27\n",
      "31\n",
      "35\n",
      "39\n",
      "43\n",
      "47\n",
      "51\n",
      "55\n",
      "59\n",
      "63\n",
      "67\n",
      "71\n",
      "75\n",
      "79\n",
      "83\n",
      "87\n",
      "91\n",
      "95\n",
      "99\n",
      "Finished Intersection constraints!\n",
      "Gurobi Optimizer version 12.0.3 build v12.0.3rc0 (linux64 - \"Rocky Linux 9.7 (Blue Onyx)\")\n",
      "\n",
      "CPU model: Intel(R) Xeon(R) Gold 6338 CPU @ 2.00GHz, instruction set [SSE2|AVX|AVX2|AVX512]\n",
      "Thread count: 64 physical cores, 128 logical processors, using up to 32 threads\n",
      "\n",
      "Academic license 2499075 - for non-commercial use only - registered to jo___@live.biu.ac.il\n",
      "Optimize a model with 10004 rows, 236389 columns and 40621943 nonzeros\n",
      "Model fingerprint: 0x8e2dbaf0\n",
      "Variable types: 0 continuous, 236389 integer (236389 binary)\n",
      "Coefficient statistics:\n",
      "  Matrix range     [1e+00, 2e+00]\n",
      "  Objective range  [8e-05, 1e+00]\n",
      "  Bounds range     [1e+00, 1e+00]\n",
      "  RHS range        [1e+00, 3e+00]\n",
      "Found heuristic solution: objective -4.3029485\n",
      "Presolve removed 159 rows and 26 columns (presolve time = 6s)...\n",
      "Presolve removed 598 rows and 29978 columns (presolve time = 11s)...\n",
      "Presolve removed 1409 rows and 29978 columns (presolve time = 17s)...\n",
      "Presolve removed 1409 rows and 30001 columns (presolve time = 20s)...\n",
      "Presolve removed 1409 rows and 30001 columns (presolve time = 25s)...\n",
      "Presolve removed 1409 rows and 30001 columns (presolve time = 31s)...\n",
      "Presolve removed 1437 rows and 30001 columns (presolve time = 35s)...\n",
      "Presolve removed 1437 rows and 30001 columns (presolve time = 40s)...\n",
      "Presolve removed 5078 rows and 30001 columns (presolve time = 45s)...\n",
      "Presolve removed 5078 rows and 30001 columns (presolve time = 51s)...\n",
      "Presolve removed 5194 rows and 30059 columns (presolve time = 55s)...\n",
      "Presolve removed 5194 rows and 30059 columns (presolve time = 60s)...\n",
      "Presolve removed 5194 rows and 30059 columns (presolve time = 65s)...\n",
      "Presolve removed 5194 rows and 30059 columns (presolve time = 71s)...\n",
      "Presolve removed 5194 rows and 30059 columns (presolve time = 75s)...\n",
      "Presolve removed 5194 rows and 30059 columns (presolve time = 80s)...\n",
      "Presolve removed 5194 rows and 30059 columns (presolve time = 85s)...\n",
      "Presolve removed 5194 rows and 30059 columns (presolve time = 90s)...\n",
      "Sparsify removed 12668811 nonzeros (95%)\n",
      "Presolve removed 5242 rows and 29238 columns\n",
      "Presolve time: 92.88s\n",
      "Presolved: 4762 rows, 207151 columns, 665418 nonzeros\n",
      "Variable types: 217 continuous, 206934 integer (206510 binary)\n",
      "Deterministic concurrent LP optimizer: primal simplex, dual simplex, and barrier\n",
      "Showing barrier log only...\n",
      "\n",
      "Root barrier log...\n",
      "\n",
      "Ordering time: 0.24s\n",
      "\n",
      "Barrier statistics:\n",
      " AA' NZ     : 3.091e+05\n",
      " Factor NZ  : 1.322e+06 (roughly 100 MB of memory)\n",
      " Factor Ops : 5.313e+08 (less than 1 second per iteration)\n",
      " Threads    : 30\n",
      "\n",
      "                  Objective                Residual\n",
      "Iter       Primal          Dual         Primal    Dual     Compl     Time\n",
      "   0  -6.25077995e+09  3.30376295e+07  6.21e+08 2.05e+00  5.24e+05    95s\n",
      "   1  -2.09730695e+09  1.83243797e+07  2.08e+08 1.43e+00  1.78e+05    95s\n",
      "   2  -8.25478976e+08  8.05031955e+06  8.20e+07 3.47e-01  7.11e+04    95s\n",
      "   3  -3.16771924e+08  3.95241745e+06  3.14e+07 8.40e-02  2.77e+04    95s\n",
      "   4  -3.74141325e+07  2.05523209e+06  3.71e+06 5.27e-03  3.32e+03    95s\n",
      "   5  -5.19824195e+06  1.54913893e+06  5.15e+05 1.62e-13  4.68e+02    95s\n",
      "   6  -1.08651533e+06  1.36138251e+06  1.08e+05 1.74e-13  1.00e+02    95s\n",
      "   7  -4.75653835e+05  1.28196234e+06  4.71e+04 1.18e-13  4.49e+01    95s\n",
      "   8  -8.48687286e+04  1.05097756e+06  8.41e+03 1.18e-13  9.45e+00    95s\n",
      "   9  -3.19680322e+04  6.27882828e+05  3.17e+03 1.33e-13  3.69e+00    95s\n",
      "  10  -1.09104325e+04  2.87365694e+05  1.08e+03 1.25e-13  1.33e+00    95s\n",
      "  11  -6.89688297e+03  2.00712466e+05  6.83e+02 1.56e-13  8.65e-01    95s\n",
      "  12  -9.49685690e+02  2.42369016e+04  9.35e+01 8.82e-14  1.08e-01    96s\n",
      "  13  -3.69351580e+02  1.11192717e+04  3.60e+01 5.22e-14  4.57e-02    96s\n",
      "  14  -1.38983178e+02  4.06687489e+03  1.32e+01 1.19e-13  1.68e-02    96s\n",
      "  15  -4.99062551e+01  2.24653514e+03  4.39e+00 1.32e-13  7.75e-03    96s\n",
      "  16  -8.02583135e+00  1.39076526e+03  2.77e-01 7.92e-14  3.51e-03    96s\n",
      "  17  -5.93028905e+00  1.13420843e+03  7.36e-02 1.27e-13  2.78e-03    96s\n",
      "  18  -5.04359319e+00  4.47698693e+02  2.08e-04 9.01e-14  1.09e-03    96s\n",
      "  19  -4.71089104e+00  2.82178861e+02  1.55e-06 1.32e-13  6.92e-04    96s\n",
      "  20  -4.35639791e+00  1.13418295e+02  8.01e-09 1.40e-13  2.84e-04    96s\n",
      "  21  -3.75623487e+00  5.18894554e+01  3.46e-10 1.07e-13  1.34e-04    96s\n",
      "  22  -2.96180405e+00  3.53153455e+01  5.82e-11 1.49e-13  9.23e-05    96s\n",
      "  23  -2.29775107e+00  3.01013882e+01  5.82e-11 1.54e-13  7.81e-05    96s\n",
      "  24  -1.73220618e+00  2.85129901e+01  2.91e-11 9.25e-14  7.29e-05    96s\n",
      "  25  -1.56935168e+00  2.62344851e+01  2.91e-11 1.61e-13  6.70e-05    96s\n",
      "  26  -1.38534742e+00  1.99133264e+01  5.82e-11 1.27e-13  5.14e-05    96s\n",
      "  27  -1.19726017e+00  1.61390870e+01  5.82e-11 1.43e-13  4.18e-05    96s\n",
      "  28  -6.77544949e-01  1.20405387e+01  5.82e-11 1.04e-13  3.07e-05    96s\n",
      "  29  -5.25624599e-01  1.08677392e+01  5.82e-11 1.12e-13  2.75e-05    96s\n",
      "  30  -5.21307326e-01  9.40681587e+00  5.82e-11 7.23e-14  2.39e-05    96s\n",
      "  31   4.56598554e-01  6.65460718e+00  5.82e-11 1.54e-13  1.49e-05    97s\n",
      "  32   8.21614523e-01  1.36725164e+00  1.46e-11 8.40e-14  1.32e-06    97s\n",
      "  33   9.31391742e-01  9.33905436e-01  5.82e-11 1.72e-13  6.06e-09    97s\n",
      "  34   9.32998017e-01  9.33004316e-01  1.83e-10 9.41e-14  1.52e-11    97s\n",
      "  35   9.33002405e-01  9.33002405e-01  2.91e-11 1.70e-13  8.65e-16    97s\n",
      "\n",
      "Barrier solved model in 35 iterations and 96.74 seconds (137.54 work units)\n",
      "Optimal objective 9.33002405e-01\n",
      "\n",
      "\n",
      "Root crossover log...\n",
      "\n",
      "    3828 DPushes remaining with DInf 0.0000000e+00                97s\n",
      "       0 DPushes remaining with DInf 0.0000000e+00                97s\n",
      "\n",
      "       0 PPushes remaining with PInf 0.0000000e+00                97s\n",
      "\n",
      "  Push phase complete: Pinf 0.0000000e+00, Dinf 2.8908896e-14     97s\n",
      "\n",
      "\n",
      "Root simplex log...\n",
      "\n",
      "Iteration    Objective       Primal Inf.    Dual Inf.      Time\n",
      "    3594    9.3300241e-01   0.000000e+00   0.000000e+00     97s\n",
      "Concurrent spin time: 0.43s (can be avoided by choosing Method=3)\n",
      "\n",
      "Solved with barrier\n",
      "    3594    9.3300241e-01   0.000000e+00   0.000000e+00     97s\n",
      "\n",
      "Root relaxation: objective 9.330024e-01, 3594 iterations, 2.63 seconds (2.58 work units)\n",
      "\n",
      "    Nodes    |    Current Node    |     Objective Bounds      |     Work\n",
      " Expl Unexpl |  Obj  Depth IntInf | Incumbent    BestBd   Gap | It/Node Time\n",
      "\n",
      "*    0     0               0       0.9330024    0.93300  0.00%     -   96s\n",
      "\n",
      "Explored 1 nodes (3594 simplex iterations) in 97.14 seconds (137.62 work units)\n",
      "Thread count was 32 (of 128 available processors)\n",
      "\n",
      "Solution count 2: 0.933002 -4.30295 \n",
      "No other solutions better than 0.933002\n",
      "\n",
      "Optimal solution found (tolerance 1.00e-04)\n",
      "Best objective 9.330024053824e-01, best bound 9.330024053824e-01, gap 0.0000%\n",
      "Protein #1 (4)\n",
      "[\"(-80, -60, 'f')\", \"(99, 123, 'r')\", \"(78, 97, 'f')\", \"(264, 282, 'r')\"]\n",
      "\n",
      "Protein #2 (4)\n",
      "[\"(-58, -34, 'f')\", \"(119, 141, 'r')\", \"(93, 115, 'f')\", \"(277, 296, 'r')\"]\n",
      "\n",
      "Protein #3 (2)\n",
      "[\"(-31, -1, 'f')\", \"(156, 174, 'r')\"]\n",
      "\n",
      "Saved summary to: Temp/var_ILP_results.csv\n",
      "Saved ILP selected primers to: Temp/var_ILP_selected_primers_ILP.csv\n",
      "Saved greedy selected primers to: Temp/var_ILP_selected_primers_greedy.csv\n",
      "Variable ILP Results DataFrame:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>protein_name</th>\n",
       "      <th>num_variants</th>\n",
       "      <th>seq_length</th>\n",
       "      <th>graph_nodes</th>\n",
       "      <th>graph_edges</th>\n",
       "      <th>graph_time_sec</th>\n",
       "      <th>greedy_path_length</th>\n",
       "      <th>ilp_num_vars</th>\n",
       "      <th>ilp_path_length</th>\n",
       "      <th>ilp_num_constraints</th>\n",
       "      <th>ilp_setup_time_sec</th>\n",
       "      <th>ilp_optimize_time_sec</th>\n",
       "      <th>ILP_primer_efficiency</th>\n",
       "      <th>ILP_solution_feasibility</th>\n",
       "      <th>greedy_solution_feasibility</th>\n",
       "      <th>greedy_primer_efficiency</th>\n",
       "      <th>greedy_time_sec</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CXDAR</td>\n",
       "      <td>3</td>\n",
       "      <td>133</td>\n",
       "      <td>3408</td>\n",
       "      <td>236389</td>\n",
       "      <td>3.029</td>\n",
       "      <td>10</td>\n",
       "      <td>236389</td>\n",
       "      <td>10</td>\n",
       "      <td>10004</td>\n",
       "      <td>88.012</td>\n",
       "      <td>98.115</td>\n",
       "      <td>0.933002</td>\n",
       "      <td>FEASIBLE</td>\n",
       "      <td>FEASIBLE</td>\n",
       "      <td>2.526991</td>\n",
       "      <td>1.067</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  protein_name  num_variants  seq_length  graph_nodes  graph_edges  \\\n",
       "0        CXDAR             3         133         3408       236389   \n",
       "\n",
       "   graph_time_sec  greedy_path_length  ilp_num_vars  ilp_path_length  \\\n",
       "0           3.029                  10        236389               10   \n",
       "\n",
       "   ilp_num_constraints  ilp_setup_time_sec  ilp_optimize_time_sec  \\\n",
       "0                10004              88.012                 98.115   \n",
       "\n",
       "   ILP_primer_efficiency ILP_solution_feasibility greedy_solution_feasibility  \\\n",
       "0               0.933002                 FEASIBLE                    FEASIBLE   \n",
       "\n",
       "   greedy_primer_efficiency  greedy_time_sec  \n",
       "0                  2.526991            1.067  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "args.merge_bins = True\n",
    "results_df, ilp_path, greedy_path  = run_var_ilp(full_sequence, mutreg_nt, protein_name,args,cfg)\n",
    "print(\"Variable ILP Results DataFrame:\")\n",
    "results_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Running PD-mul-ILP\n",
    "\n",
    "We will run PD-mul-ILP on multiple non-homologous proteins. For this part, we will add another partial 100-nt sequence of the SHP2 protein"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "protein_name1 = 'CXDAR'\n",
    "mutreg_nt1 = 'ATGGCGCTCCTGCTGTGCTTCGTGCTCCTGTGCGGAGTAGTGGATTTCGCCAGAAGTTTGAGTATCACTACTCCTGAAGAGATGATTGAAAAAGCCAAAG'\n",
    "## Add upstream and downstream sequences\n",
    "full_sequence1 = cfg.upstream + mutreg_nt1 + cfg.downstream\n",
    "\n",
    "protein_name2 = 'SHP2'\n",
    "mutreg_nt2 = 'ATGACATCGCGGAGATGGTTTCACCCAAATATCACTGGTGTGGAGGCAGAAAACCTACTGTTGACAAGAGGAGTTGATGGCAGTTTTTTGGCAAGGCCTA'\n",
    "full_sequence2 = cfg.upstream + mutreg_nt2 + cfg.downstream\n",
    "\n",
    "## create protein dataset\n",
    "mutreg_regions = [mutreg_nt1, mutreg_nt2]\n",
    "sequences_nt = [full_sequence1, full_sequence2]\n",
    "protein_names = [protein_name1, protein_name2]\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Run PD-mul-ILP 2 non-homolous proteins parts\n",
    "PD-mul-ILP uses an integer linear program with forbidden pair constraints on all cross-hybridizing pairs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[STEP] Creating graphs for 2 proteins...\n",
      "Creating Graph for protein:  0 Protein name:  CXDAR\n",
      "Creating Graph for protein:  1 Protein name:  SHP2\n",
      "[DONE] Graphs created in 19.00 sec (peak 133.5 MB).\n",
      "[STEP] Finding forbidden pairs across proteins...\n",
      "Finding intra-protein forbidden pairs constraints\n",
      "[1/2] Processing CXDAR...\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[2/2] Processing SHP2...\n",
      "Number of intra-protein forbidden pairs constraints:  0\n",
      "Finding inter-protein forbidden pairs for 1 combinations...\n",
      "[1/1] Fractions of pairs considered\n",
      "Number of inter-protein forbidden pairs constraints:  0\n",
      "[DONE] Forbidden pairs in 24.56 sec (intra=0, inter=0).\n",
      "[STEP] Running ILP model...\n",
      "Creating ILP model...\n",
      "Set parameter WLSAccessID\n",
      "Set parameter WLSSecret\n",
      "Set parameter LicenseID to value 2499075\n",
      "Academic license 2499075 - for non-commercial use only - registered to jo___@live.biu.ac.il\n",
      "Adding constraints to ILP model...\n",
      "Gurobi Optimizer version 12.0.3 build v12.0.3rc0 (linux64 - \"Rocky Linux 9.7 (Blue Onyx)\")\n",
      "\n",
      "CPU model: Intel(R) Xeon(R) Gold 6338 CPU @ 2.00GHz, instruction set [SSE2|AVX|AVX2|AVX512]\n",
      "Thread count: 64 physical cores, 128 logical processors, using up to 32 threads\n",
      "\n",
      "Academic license 2499075 - for non-commercial use only - registered to jo___@live.biu.ac.il\n",
      "Optimize a model with 5100 rows, 321132 columns and 642264 nonzeros\n",
      "Model fingerprint: 0xd0d16cd9\n",
      "Variable types: 0 continuous, 321132 integer (321132 binary)\n",
      "Coefficient statistics:\n",
      "  Matrix range     [1e+00, 1e+00]\n",
      "  Objective range  [2e-04, 1e+00]\n",
      "  Bounds range     [1e+00, 1e+00]\n",
      "  RHS range        [1e+00, 1e+00]\n",
      "Found heuristic solution: objective -2.7100475\n",
      "Presolve removed 4212 rows and 308202 columns\n",
      "Presolve time: 0.78s\n",
      "Presolved: 888 rows, 12930 columns, 25860 nonzeros\n",
      "Found heuristic solution: objective 1.8129972\n",
      "Variable types: 0 continuous, 12930 integer (12930 binary)\n",
      "Performing another presolve...\n",
      "Presolve removed 888 rows and 12930 columns\n",
      "Presolve time: 0.02s\n",
      "Found heuristic solution: objective 3.3758909\n",
      "\n",
      "Explored 1 nodes (0 simplex iterations) in 0.90 seconds (2.02 work units)\n",
      "Thread count was 32 (of 128 available processors)\n",
      "\n",
      "Solution count 3: 3.37589 1.813 -2.71005 \n",
      "\n",
      "Optimal solution found (tolerance 1.00e-04)\n",
      "Best objective 3.375890943249e+00, best bound 3.375890943249e+00, gap 0.0000%\n",
      "Protein #CXDAR (4)\n",
      "[\"(-80, -60, 'f')\", \"(96, 115, 'r')\", \"(70, 96, 'f')\", \"(244, 265, 'r')\"]\n",
      "\n",
      "Protein #SHP2 (4)\n",
      "[\"(-80, -60, 'f')\", \"(98, 118, 'r')\", \"(69, 92, 'f')\", \"(244, 264, 'r')\"]\n",
      "\n",
      "[DONE] ILP status=2, objective=3.38, time=1.12 sec.\n",
      "[SAVE] Appended summary → Temp/mul_ilp_results.csv\n",
      "[SAVE] Primer paths CSV → Temp/mul_ilp_selected_primers.csv\n",
      "PD-mul_ILP Results DataFrame:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>num_proteins</th>\n",
       "      <th>graph_time_sec</th>\n",
       "      <th>ilp_num_vars</th>\n",
       "      <th>ilp_num_constraints</th>\n",
       "      <th>ilp_intra_forbidden_cnt</th>\n",
       "      <th>ilp_inter_forbidden_cnt</th>\n",
       "      <th>forbidden_time_sec</th>\n",
       "      <th>ilp_setup_time_sec</th>\n",
       "      <th>ilp_optimize_time_sec</th>\n",
       "      <th>ilp_feasibility</th>\n",
       "      <th>total_primer_efficiency</th>\n",
       "      <th>num_primers</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2</td>\n",
       "      <td>19</td>\n",
       "      <td>321132</td>\n",
       "      <td>5100</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>24.560111</td>\n",
       "      <td>12.352175</td>\n",
       "      <td>1.121767</td>\n",
       "      <td>FEASIBLE</td>\n",
       "      <td>3.375891</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   num_proteins  graph_time_sec  ilp_num_vars  ilp_num_constraints  \\\n",
       "0             2              19        321132                 5100   \n",
       "\n",
       "   ilp_intra_forbidden_cnt  ilp_inter_forbidden_cnt  forbidden_time_sec  \\\n",
       "0                        0                        0           24.560111   \n",
       "\n",
       "   ilp_setup_time_sec  ilp_optimize_time_sec ilp_feasibility  \\\n",
       "0           12.352175               1.121767        FEASIBLE   \n",
       "\n",
       "   total_primer_efficiency  num_primers  \n",
       "0                 3.375891            8  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "run_df, primer_paths = run_mul_ilp(mutreg_regions, sequences_nt, protein_names, args, cfg)\n",
    "\n",
    "print(\"PD-mul_ILP Results DataFrame:\")\n",
    "run_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Primer slection and total efficiency\n",
    "\n",
    "We will look at the primer paths found for each protein by PD-mul-ILP, and the total efficiency of the selected primers."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Chosen primers:\n",
      "Protein: CXDAR\n",
      "[\"(-80, -60, 'f')\", \"(96, 115, 'r')\", \"(70, 96, 'f')\", \"(244, 265, 'r')\"]\n",
      "Protein: SHP2\n",
      "[\"(-80, -60, 'f')\", \"(98, 118, 'r')\", \"(69, 92, 'f')\", \"(244, 264, 'r')\"]\n",
      "Total primer efficiency:\n",
      "3.3758909432490496\n"
     ]
    }
   ],
   "source": [
    "print(\"Chosen primers:\")\n",
    "\n",
    "for protein, path in primer_paths.items():\n",
    "    print(\"Protein:\", protein)\n",
    "    print(path)\n",
    "\n",
    "total_efficiency = run_df['total_primer_efficiency'].prod()   \n",
    "print(\"Total primer efficiency:\")\n",
    "print(total_efficiency) "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Testing behavior with infeasible primer sequences\n",
    "\n",
    "We test a case where no feasible primer set exists. Specifically, two sequences are constructed such that their upstream regions are reverse complements, creating primer pairs with high melting temperature and cross-hybridization risk.\n",
    "\n",
    "These primer pairs are forbidden by the ILP constraints, and therefore PrimerDesigner correctly reports that no feasible solution exists."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "cfg_mod = GU.Config(\n",
    "    upstream='ATGGCGCTCCTGCTGTGC',\n",
    "    downstream=cfg.downstream,\n",
    "    max_tm=cfg.max_tm\n",
    ")\n",
    "\n",
    "args.primer_lmax = 18\n",
    "\n",
    "protein_name1 = 'CXDAR'\n",
    "mutreg_nt1 = 'ATGGCGCTCCTGCTGTGCTTCGTGCTCCTGTGCGGAGTAGTGGATTTCGCCAGAAGTTTGAGTATCACTACTCCTGAAGAGATGATTGAAAAAGCCAAAG'\n",
    "## Add upstream and downstream sequences\n",
    "full_sequence1 = cfg_mod.upstream + mutreg_nt1 + cfg.downstream\n",
    "\n",
    "## Add reverse complement upstream sequence for the other protein\n",
    "protein_name2 = 'SHP2'\n",
    "mutreg_nt2 = 'ATGACATCGCGGAGATGGTTTCACCCAAATATCACTGGTGTGGAGGCAGAAAACCTACTGTTGACAAGAGGAGTTGATGGCAGTTTTTTGGCAAGGCCTA'\n",
    "# full_sequence2 = cfg_mod.upstream + mutreg_nt2 + cfg.downstream\n",
    "full_sequence2 = GU.revcomp(cfg_mod.upstream) + mutreg_nt2 + cfg.downstream\n",
    "\n",
    "## create protein dataset\n",
    "mutreg_regions = [mutreg_nt1, mutreg_nt2]\n",
    "sequences_nt = [full_sequence1, full_sequence2]\n",
    "protein_names = [protein_name1, protein_name2]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[STEP] Creating graphs for 2 proteins...\n",
      "Creating Graph for protein:  0 Protein name:  CXDAR\n",
      "Creating Graph for protein:  1 Protein name:  SHP2\n",
      "[DONE] Graphs created in 0.00 sec (peak 1.6 MB).\n",
      "[STEP] Finding forbidden pairs across proteins...\n",
      "Finding intra-protein forbidden pairs constraints\n",
      "[1/2] Processing CXDAR...\n",
      "[2/2] Processing SHP2...\n",
      "Number of intra-protein forbidden pairs constraints:  0\n",
      "Finding inter-protein forbidden pairs for 1 combinations...\n",
      "[1/1] Fractions of pairs considered\n",
      "Number of inter-protein forbidden pairs constraints:  100\n",
      "[DONE] Forbidden pairs in 7.50 sec (intra=0, inter=100).\n",
      "[STEP] Running ILP model...\n",
      "Creating ILP model...\n",
      "Set parameter WLSAccessID\n",
      "Set parameter WLSSecret\n",
      "Set parameter LicenseID to value 2499075\n",
      "Academic license 2499075 - for non-commercial use only - registered to jo___@live.biu.ac.il\n",
      "Adding constraints to ILP model...\n",
      "Gurobi Optimizer version 12.0.3 build v12.0.3rc0 (linux64 - \"Rocky Linux 9.7 (Blue Onyx)\")\n",
      "\n",
      "CPU model: Intel(R) Xeon(R) Gold 6338 CPU @ 2.00GHz, instruction set [SSE2|AVX|AVX2|AVX512]\n",
      "Thread count: 64 physical cores, 128 logical processors, using up to 32 threads\n",
      "\n",
      "Academic license 2499075 - for non-commercial use only - registered to jo___@live.biu.ac.il\n",
      "Optimize a model with 128 rows, 46 columns and 389 nonzeros\n",
      "Model fingerprint: 0x45f747dd\n",
      "Variable types: 0 continuous, 46 integer (46 binary)\n",
      "Coefficient statistics:\n",
      "  Matrix range     [1e+00, 1e+00]\n",
      "  Objective range  [1e-02, 3e-01]\n",
      "  Bounds range     [1e+00, 1e+00]\n",
      "  RHS range        [1e+00, 1e+00]\n",
      "Presolve removed 98 rows and 44 columns\n",
      "Presolve time: 0.00s\n",
      "\n",
      "Explored 0 nodes (0 simplex iterations) in 0.01 seconds (0.00 work units)\n",
      "Thread count was 1 (of 128 available processors)\n",
      "\n",
      "Solution count 0\n",
      "No other solutions better than -1e+100\n",
      "\n",
      "Model is infeasible\n",
      "Best objective -, best bound -, gap -\n",
      "Model status: 3\n",
      "[DONE] ILP status=3, objective=nan, time=0.03 sec.\n",
      "[SAVE] Appended summary → Temp/mul_ilp_results.csv\n",
      "[SAVE] Primer paths CSV → Temp/mul_ilp_selected_primers.csv\n",
      "PD-mul_ILP Results DataFrame:\n"
     ]
    }
   ],
   "source": [
    "run_df, primer_paths = run_mul_ilp(mutreg_regions, sequences_nt, protein_names, args, cfg_mod)\n",
    "\n",
    "print(\"PD-mul_ILP Results DataFrame:\")\n",
    "run_df\n",
    "\n",
    "## reseet primer_lmax\n",
    "args.primer_lmax = 30"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Running PD-mul-Greedy\n",
    "\n",
    "We will run PD-mul-Greedy, which uses a greedy algorithm to find efficient primers paths for each protein, on multiple non-homologous proteins. \n",
    "We will use the same proteins as we used for the PD-mul-ILP run. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[INFO] Processing protein 0/2: CXDAR\n",
      "[SAVE] Selected primers CSV → Temp/PD_mul_Greedy_selected_primers.csv\n",
      "PD-mul_greedy Results DataFrame:\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{'num_proteins': 2,\n",
       " 'total_primer_efficiency': 2.329953557394205,\n",
       " 'greedy_time_sec': 10.817,\n",
       " 'cross_hybridizations_cnt': 0,\n",
       " 'proteins_with_reiterations_cnt': 0,\n",
       " 'total_reiterations': 0,\n",
       " 'unresolved_proteins_cnt': 0,\n",
       " 'unresolved_proteins': '',\n",
       " 'total_primers': 4}"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "run_df, paths = run_mul_greedy(sequences_nt,mutreg_regions, protein_names, args, cfg)\n",
    "\n",
    "print(\"PD-mul_greedy Results DataFrame:\")\n",
    "run_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Primers found for each protein total efficiency\n",
    "\n",
    "We will examine the total efficiency and the primers found for each of the proteins by the greedy algorithm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Protein: CXDAR\n",
      "[(-51, -33, 'f'), (125, 146, 'r')]\n",
      "Protein: SHP2\n",
      "[(-74, -54, 'f'), (113, 131, 'r')]\n",
      "Total primer efficiency:\n",
      "2.329953557394205\n"
     ]
    }
   ],
   "source": [
    "for protein, path in paths.items():\n",
    "    print(\"Protein:\", protein)\n",
    "    print(path)\n",
    "\n",
    "total_efficiency = run_df['total_primer_efficiency']   \n",
    "print(\"Total primer efficiency:\")\n",
    "print(total_efficiency)\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "primer_env3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.25"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
