{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### All Imports"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from General.args import *\n",
    "import General.utils as GU\n",
    "from PD_var_ILP.run_var_ilp import *\n",
    "from PD_mul_ILP.run_mul_ilp import *\n",
    "from PD_mul_greedy.run_mul_greedy import *\n",
    "from PD_single_LPath.run_pd_single_LPath import *"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Set up gurobi optimization library\n",
    "\n",
    "\n",
    "First, create a file called gurobi.json with the details of your gurobi license in the main directory. The file should be in the following format:\n",
    "\n",
    "```json\n",
    "{\n",
    "  \"WLSACCESSID\": \"XXXXX\",\n",
    "  \"WLSSECRET\": \"XXXXX\",\n",
    "  \"LICENSEID\": 12345\n",
    "}\n",
    "```\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Global program cmd arguments\n",
    "\n",
    "Set arguments to their default values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "args = get_args()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Define global variables for upstream and downstream regions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "cfg = load_config(\"config.json\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Run PD-Single-LPath"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Define sequence\n",
    "First, we will define our mutreg region sequence, we will take the first 100 nucleotides of the coding sequence of the CXDAR human protein gene as a small test dataset."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "protein_name = 'CXDAR'\n",
    "mutreg_nt = 'ATGGCGCTCCTGCTGTGCTTCGTGCTCCTGTGCGGAGTAGTGGATTTCGCCAGAAGTTTGAGTATCACTACTCCTGAAGAGATGATTGAAAAAGCCAAAG'\n",
    "\n",
    "## Add upstream and downstream sequences\n",
    "full_sequence = cfg.upstream + mutreg_nt + cfg.downstream"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Run longest path algorithm\n",
    "\n",
    "We will runs PD-Single_LPath which runs a longest path algorithms on the primer graph, which is a DAG, to find the most efficient primers."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Creating primer df\n",
      "Creating graph\n",
      "Saved CSV summary to: results/PD_single_LPath_results.csv\n",
      "Saved path details to: results/PD_single_LPath_selected_primers.json\n",
      "Results DataFrame:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>protein_name</th>\n",
       "      <th>graph_nodes</th>\n",
       "      <th>graph_edges</th>\n",
       "      <th>graph_time_sec</th>\n",
       "      <th>longest_path_efficiency</th>\n",
       "      <th>total_time_sec</th>\n",
       "      <th>num_primers</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CXDAR</td>\n",
       "      <td>2550</td>\n",
       "      <td>160566</td>\n",
       "      <td>2.168</td>\n",
       "      <td>1.511491</td>\n",
       "      <td>7.416</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  protein_name  graph_nodes  graph_edges  graph_time_sec  \\\n",
       "0        CXDAR         2550       160566           2.168   \n",
       "\n",
       "   longest_path_efficiency  total_time_sec  num_primers  \n",
       "0                 1.511491           7.416            4  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "run_df, primer_paths = run_longest_path(full_sequence, mutreg_nt, protein_name, args, cfg) \n",
    "\n",
    "print(\"Results DataFrame:\")\n",
    "run_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Selected primer path and primer efficiency\n",
    "\n",
    "Print all selected primers in longest path and their efficiency"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Primer Paths:\n",
      "(-80, -60, 'f')\n",
      "(96, 115, 'r')\n",
      "(70, 96, 'f')\n",
      "(244, 265, 'r')\n",
      "Primer effiency\n",
      "1.5114909044453393\n"
     ]
    }
   ],
   "source": [
    "print(\"Primer Paths:\")\n",
    "for path in primer_paths:\n",
    "    print(path)\n",
    "\n",
    "print(\"Primer effiency\")\n",
    "print(run_df['longest_path_efficiency'].prod())   "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Run PD-var-ILP\n",
    "\n",
    "We will run PD-var-ILP on multiple variants of the same protein.\n",
    "\n",
    "For this part, we will use the same sequence of the CXDAR protein as PD-single-LPath.\n",
    "We will use the default number of variants, which is 3, and the deafult maximum overlap length, which is 6. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Creating primer df\n",
      "Creating graph\n",
      "Running greedy algorithm\n",
      "Running ILP\n",
      "Set parameter WLSAccessID\n",
      "Set parameter WLSSecret\n",
      "Set parameter LicenseID to value 2499075\n",
      "Academic license 2499075 - for non-commercial use only - registered to jo___@live.biu.ac.il\n",
      "Number of Constraints: 5520\n",
      "Average Vars Per Constraint 80.94384057971014\n",
      "Graph edges:  160566\n",
      "Finished ILP Variable Creations\n",
      "0\n",
      "3\n",
      "7\n",
      "11\n",
      "15\n",
      "19\n",
      "23\n",
      "27\n",
      "31\n",
      "35\n",
      "39\n",
      "43\n"
     ]
    }
   ],
   "source": [
    "results_df, ilp_path, greedy_path  = run_var_ilp(full_sequence, mutreg_nt, protein_name,args,cfg)\n",
    "print(\"Variable ILP Results DataFrame:\")\n",
    "results_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Compare PD-Var-ILP with greedy baseline\n",
    "\n",
    "We will compare the primers found by the ILP and by the greedy approach and their total efficiencies."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ILP chosen primers:\n",
      "Variant 1\n",
      "[\"(-86, -68, 'f')\", \"(87, 109, 'r')\", \"(59, 85, 'f')\", \"(244, 263, 'r')\"]\n",
      "Variant 2\n",
      "[\"(-74, -54, 'f')\", \"(106, 124, 'r')\"]\n",
      "Variant 3\n",
      "[\"(-58, -34, 'f')\", \"(123, 141, 'r')\"]\n",
      "ILP primer efficiency:  2.4237929216219447\n",
      "Greedy chosen primers:\n",
      "Variant 1\n",
      "[(-80, -60, 'f'), (96, 115, 'r'), (70, 96, 'f'), (244, 265, 'r')]\n",
      "Variant 2\n",
      "[(-58, -34, 'f'), (123, 141, 'r')]\n",
      "Variant 3\n",
      "[(-39, -9, 'f'), (136, 156, 'r')]\n",
      "Greedy primer efficiency:  1.514172485959763\n"
     ]
    }
   ],
   "source": [
    "print(\"ILP chosen primers:\")\n",
    "\n",
    "for i,path in enumerate(ilp_path):\n",
    "    print(\"Variant\", i+1)\n",
    "    print(path)\n",
    "\n",
    "print(\"ILP primer efficiency: \",results_df['ILP_primer_efficiency'].prod())\n",
    "\n",
    "\n",
    "print(\"Greedy chosen primers:\")\n",
    "for i,path in enumerate(greedy_path):\n",
    "    print(\"Variant\", i+1)\n",
    "    print(path)\n",
    "\n",
    "print(\"Greedy primer efficiency: \",results_df['greedy_primer_efficiency'].prod())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Running PD-mul-ILP\n",
    "\n",
    "We will run PD-mul-ILP on multiple non-homologous proteins. For this part, we will add another partial 100-nt sequence of the SHP2 protein"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "protein_name1 = 'CXDAR'\n",
    "mutreg_nt1 = 'ATGGCGCTCCTGCTGTGCTTCGTGCTCCTGTGCGGAGTAGTGGATTTCGCCAGAAGTTTGAGTATCACTACTCCTGAAGAGATGATTGAAAAAGCCAAAG'\n",
    "## Add upstream and downstream sequences\n",
    "full_sequence1 = cfg.upstream + mutreg_nt1 + cfg.downstream\n",
    "\n",
    "protein_name2 = 'SHP2'\n",
    "mutreg_nt2 = 'ATGACATCGCGGAGATGGTTTCACCCAAATATCACTGGTGTGGAGGCAGAAAACCTACTGTTGACAAGAGGAGTTGATGGCAGTTTTTTGGCAAGGCCTA'\n",
    "full_sequence2 = cfg.upstream + mutreg_nt2 + cfg.downstream\n",
    "\n",
    "## create protein dataset\n",
    "mutreg_regions = [mutreg_nt1, mutreg_nt2]\n",
    "sequences_nt = [full_sequence1, full_sequence2]\n",
    "protein_names = [protein_name1, protein_name2]\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Run PD-mul-ILP 2 non-homolous proteins parts\n",
    "PD-mul-ILP uses an integer linear program with forbidden pair constraints on all cross-hybridizing pairs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[STEP] Creating graphs for 2 proteins...\n",
      "Creating Graph for protein:  0 Protein name:  CXDAR\n",
      "Creating primer df\n",
      "Creating graph\n",
      "Creating Graph for protein:  1 Protein name:  SHP2\n",
      "Creating primer df\n",
      "Creating graph\n",
      "[DONE] Graphs created in 23.00 sec (peak 136.6 MB).\n",
      "[STEP] Finding forbidden pairs across proteins...\n",
      "Finding intra-protein forbidden pairs constraints\n",
      "[1/2] Processing CXDAR...\n",
      "[2/2] Processing SHP2...\n",
      "Number of intra-protein forbidden pairs constraints:  0\n",
      "Finding inter-protein forbidden pairs for 1 combinations...\n",
      "[1/1] Fractions of pairs considered\n"
     ]
    },
    {
     "ename": "TypeError",
     "evalue": "find_forbidden_pairs_inter() missing 1 required positional argument: 'cfg'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[20], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m run_df, primer_paths \u001b[38;5;241m=\u001b[39m \u001b[43mrun_mul_ilp\u001b[49m\u001b[43m(\u001b[49m\u001b[43mmutreg_regions\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43msequences_nt\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mprotein_names\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mcfg\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m      3\u001b[0m \u001b[38;5;28mprint\u001b[39m(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mPD-mul_ILP Results DataFrame:\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m      4\u001b[0m run_df\n",
      "File \u001b[0;32m/localdata/jonathan/PrimerDesigner/PD_mul_ILP/run_mul_ilp.py:34\u001b[0m, in \u001b[0;36mrun_mul_ilp\u001b[0;34m(mutreg_regions, sequences_nt, protein_names, args, cfg)\u001b[0m\n\u001b[1;32m     32\u001b[0m \u001b[38;5;28mprint\u001b[39m(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m[STEP] Finding forbidden pairs across proteins...\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m     33\u001b[0m t_forbid \u001b[38;5;241m=\u001b[39m time\u001b[38;5;241m.\u001b[39mtime()\n\u001b[0;32m---> 34\u001b[0m single_forbidden, multiple_forbidden, single_pair_cnt, multi_pairs_cnt \u001b[38;5;241m=\u001b[39m \u001b[43mfind_forbidden_pairs\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m     35\u001b[0m \u001b[43m    \u001b[49m\u001b[43mprotein_names\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43msequences_nt\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mcfg\u001b[49m\n\u001b[1;32m     36\u001b[0m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m     37\u001b[0m forbidden_time \u001b[38;5;241m=\u001b[39m time\u001b[38;5;241m.\u001b[39mtime() \u001b[38;5;241m-\u001b[39m t_forbid\n\u001b[1;32m     38\u001b[0m \u001b[38;5;28mprint\u001b[39m(\u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m[DONE] Forbidden pairs in \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mforbidden_time\u001b[38;5;132;01m:\u001b[39;00m\u001b[38;5;124m.2f\u001b[39m\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m sec \u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m     39\u001b[0m       \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m(intra=\u001b[39m\u001b[38;5;132;01m{\u001b[39;00msingle_pair_cnt\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m, inter=\u001b[39m\u001b[38;5;132;01m{\u001b[39;00mmulti_pairs_cnt\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m).\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n",
      "File \u001b[0;32m/localdata/jonathan/PrimerDesigner/PD_mul_ILP/ilp_model.py:65\u001b[0m, in \u001b[0;36mfind_forbidden_pairs\u001b[0;34m(protein_names, sequences_nt, args, cfg)\u001b[0m\n\u001b[1;32m     62\u001b[0m protein2 \u001b[38;5;241m=\u001b[39m protein_names[p2]\n\u001b[1;32m     64\u001b[0m \u001b[38;5;66;03m# Find forbidden pairs between two protein sequences\u001b[39;00m\n\u001b[0;32m---> 65\u001b[0m forbidden_pairs \u001b[38;5;241m=\u001b[39m \u001b[43mfind_forbidden_pairs_inter\u001b[49m\u001b[43m(\u001b[49m\u001b[43msequence1\u001b[49m\u001b[43m,\u001b[49m\u001b[43msequence2\u001b[49m\u001b[43m,\u001b[49m\u001b[43margs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m     67\u001b[0m multiple_forbidden[(protein1,protein2)]\u001b[38;5;241m=\u001b[39m forbidden_pairs\n\u001b[1;32m     69\u001b[0m multiple_forbidden_cnt \u001b[38;5;241m+\u001b[39m\u001b[38;5;241m=\u001b[39m \u001b[38;5;28mlen\u001b[39m(forbidden_pairs)\n",
      "\u001b[0;31mTypeError\u001b[0m: find_forbidden_pairs_inter() missing 1 required positional argument: 'cfg'"
     ]
    }
   ],
   "source": [
    "\n",
    "run_df, primer_paths = run_mul_ilp(mutreg_regions, sequences_nt, protein_names, args, cfg)\n",
    "\n",
    "print(\"PD-mul_ILP Results DataFrame:\")\n",
    "run_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Primer slection and total efficiency\n",
    "\n",
    "We will look at the primer paths found for each protein by PD-mul-ILP, and the total efficiency of the selected primers."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Chosen primers:\n",
      "Protein: CXDAR\n",
      "[\"(-80, -60, 'f')\", \"(96, 115, 'r')\", \"(70, 96, 'f')\", \"(244, 265, 'r')\"]\n",
      "Protein: SHP2\n",
      "[\"(-80, -60, 'f')\", \"(98, 118, 'r')\", \"(69, 92, 'f')\", \"(244, 264, 'r')\"]\n",
      "Total primer efficiency:\n",
      "3.3758909432490496\n"
     ]
    }
   ],
   "source": [
    "print(\"Chosen primers:\")\n",
    "\n",
    "for protein, path in primer_paths.items():\n",
    "    print(\"Protein:\", protein)\n",
    "    print(path)\n",
    "\n",
    "total_efficiency = run_df['total_primer_efficiency'].prod()   \n",
    "print(\"Total primer efficiency:\")\n",
    "print(total_efficiency) "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Testing behaviour with infeasible sequences\n",
    "We will test what happens when upstream region on the 2 sequences reverse complements and the 2 primers are forbidden pairs. No feasible solution should be found in this case"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "UPSTREAM  = 'ATGGCGCTCCTGCTGTGC'\n",
    "args.primer_lmax = 18\n",
    "\n",
    "protein_name1 = 'CXDAR'\n",
    "mutreg_nt1 = 'ATGGCGCTCCTGCTGTGCTTCGTGCTCCTGTGCGGAGTAGTGGATTTCGCCAGAAGTTTGAGTATCACTACTCCTGAAGAGATGATTGAAAAAGCCAAAG'\n",
    "## Add upstream and downstream sequences\n",
    "full_sequence1 = cfg.upstream + mutreg_nt1 + cfg.downstream\n",
    "\n",
    "## Add reverse complement upstream sequence for the other protein\n",
    "protein_name2 = 'SHP2'\n",
    "mutreg_nt2 = 'ATGACATCGCGGAGATGGTTTCACCCAAATATCACTGGTGTGGAGGCAGAAAACCTACTGTTGACAAGAGGAGTTGATGGCAGTTTTTTGGCAAGGCCTA'\n",
    "full_sequence2 = GU.revcomp(cfg.upstream) + mutreg_nt2 + cfg.downstream\n",
    "\n",
    "## create protein dataset\n",
    "mutreg_regions = [mutreg_nt1, mutreg_nt2]\n",
    "sequences_nt = [full_sequence1, full_sequence2]\n",
    "protein_names = [protein_name1, protein_name2]\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Running PD-mul-Greedy\n",
    "\n",
    "We will run PD-mul-Greedy, which uses a greedy algorithm to find efficient primers paths for each protein, on multiple non-homologous proteins. \n",
    "We will use the same proteins as we used for the PD-mul-ILP run. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[INFO] Processing protein 0/2: CXDAR\n",
      "Creating primer df\n",
      "Creating graph\n",
      "Creating primer df\n",
      "Creating graph\n",
      " Saved summary to: Test_outputs/summary.csv\n",
      " Saved per-protein metrics to: Test_outputs/per_protein_metrics.csv\n",
      " Saved paths to: Test_outputs/primers_per_protein.json\n",
      "PD-mul_greedy Results DataFrame:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>num_proteins</th>\n",
       "      <th>total_primer_efficiency</th>\n",
       "      <th>greedy_time_sec</th>\n",
       "      <th>cross_hybridizations_cnt</th>\n",
       "      <th>proteins_with_reiterations_cnt</th>\n",
       "      <th>total_reiterations</th>\n",
       "      <th>unresolved_proteins_cnt</th>\n",
       "      <th>unresolved_proteins</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2</td>\n",
       "      <td>3.375891</td>\n",
       "      <td>13.527</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   num_proteins  total_primer_efficiency  greedy_time_sec  \\\n",
       "0             2                 3.375891           13.527   \n",
       "\n",
       "   cross_hybridizations_cnt  proteins_with_reiterations_cnt  \\\n",
       "0                         0                               0   \n",
       "\n",
       "   total_reiterations  unresolved_proteins_cnt unresolved_proteins  \n",
       "0                   0                        0                      "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "run_df, paths = run_mul_greedy(sequences_nt,mutreg_regions, protein_names, args, cfg)\n",
    "\n",
    "print(\"PD-mul_greedy Results DataFrame:\")\n",
    "run_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Primers found for each protein total efficiency\n",
    "\n",
    "We will examine the total efficiency and the primers found for each of the proteins by the greedy algorithm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Protein: CXDAR\n",
      "[(-80, -60, 'f'), (96, 115, 'r'), (70, 96, 'f'), (244, 265, 'r')]\n",
      "Protein: SHP2\n",
      "[(-80, -60, 'f'), (98, 118, 'r'), (69, 92, 'f'), (244, 264, 'r')]\n",
      "Total primer efficiency:\n",
      "3.3758909432490496\n"
     ]
    }
   ],
   "source": [
    "for protein, path in paths.items():\n",
    "    print(\"Protein:\", protein)\n",
    "    print(path)\n",
    "\n",
    "total_efficiency = run_df['total_primer_efficiency'].prod()   \n",
    "print(\"Total primer efficiency:\")\n",
    "print(total_efficiency)\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "primer_env3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.25"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
